cores/arduino/HardwareTimer.cpp

```diff
@@ -660,7 +632,11 @@ void HardwareTimer::setInterruptPriority(uint32_t preemptPriority, uint32_t subP
   */
 void HardwareTimer::attachInterrupt(void (*callback)(HardwareTimer *))
 {
-  callbacks[0] = callback;
+  if (callbacks[0] == NULL) { //if there's no callback the ISR here is not enabled
+    __HAL_TIM_ENABLE_IT(&(_timerObj.handle), TIM_IT_UPDATE);
```

Comment:
There are some use case were this will not work:
__HAL_TIM_ENABLE_IT() will modify hardware register to enable interrupt, but in the following example case, 
```
  MyTim->setMode(channel, TIMER_OUTPUT_COMPARE_PWM1, pin);
  // MyTim->setPrescaleFactor(8); // Due to setOverflow with MICROSEC_FORMAT, prescaler will be computed automatically based on timer input clock
  MyTim->setOverflow(100000, MICROSEC_FORMAT); // 10000 microseconds = 10 milliseconds
  MyTim->setCaptureCompare(channel, 50, PERCENT_COMPARE_FORMAT); // 50%
  MyTim->attachInterrupt(Update_IT_callback);
  MyTim->attachInterrupt(channel, Compare_IT_callback);
  MyTim->resume();
```
the timer peripheral clock (the clock of register) is enabled only on attachInterrupt() (thanks to HAL_TIM_Base_Init()). Thus it is not possible to write register on __HAL_TIM_ENABLE_IT().
For information, I inherit this behavior, from danieleff implementation (see https://github.com/stm32duino/Arduino_Core_STM32/issues/146#issuecomment-503980512)

Reply:
Sorry @ABOSTM I think I didn't understood....Can't you write a register if the timer is disabled? What prevents it?
The only thing which could make this not work is that HAL_TIM_Base_Init overwrites the DIER register....but from what I saw (before making the changes and the PR) it doesn't... TIM_Base_SetConfig doesn't touch that register...
So actually you set the DIER register in the attachInterrupt and then you start the timer....which should behave accordingly to what the registers say, generating the interrupt.

What am I missing?

Reply:
@LinoBarreca , the point is to provide the clock of let say TIM1 peripheral. Without this clock, register are not accessible.  This is done by writing some RCC (Reset and Clock Controller). This is currently done thanks to:
HAL_TIM_Base_MspInit() --> enableTimerClock()  from timer.c
And this is called from HAL_TIM_Base_Init().